src/interfaces/coral_web/src/hooks/session.ts

```diff
@@ -0,0 +1,86 @@
+import { useLocalStorageValue } from '@react-hookz/web';
+import { useMutation } from '@tanstack/react-query';
+import { jwtDecode } from 'jwt-decode';
+import { useRouter } from 'next/router';
+import { useCallback, useMemo } from 'react';
+
+import { useCohereClient } from '@/cohere-client';
+import { LOCAL_STORAGE_KEYS } from '@/constants';
+import { useServerAuthStrategies } from '@/hooks/authStrategies';
+
+interface LoginParams {
+  email: string;
+  password: string;
+}
+
+interface RegisterParams {
+  name: string;
+  email: string;
+  password: string;
+}
+
+interface UserSession {
+  email: string;
+  fullname: string;
+  id: string;
+}
+
+export const useSession = () => {
+  const router = useRouter();
+  const { data: authStrategies } = useServerAuthStrategies();
+  const {
+    value: authToken,
+    set: setAuthToken,
+    remove: clearAuthToken,
+  } = useLocalStorageValue<string | undefined>(LOCAL_STORAGE_KEYS.authToken, {
+    defaultValue: undefined,
+  });
+
+  const isLoggedIn = useMemo(
+    () =>
+      (authStrategies && authStrategies.length > 0 && !!authToken) ||
+      !authStrategies ||
+      authStrategies.length === 0,
+    [authToken, authStrategies]
+  );
+
+  const cohereClient = useCohereClient();
+  const session: UserSession = authToken ? jwtDecode(authToken).context : {};
+
+  const loginMutation = useMutation({
```

Comment:
could we extract out these mutations and then use them in here like
```
const { mutateAsync: login } = useLogin()
...
const handleLogin = async () => {
  try {
    const res = await login(params);
    setAuthToken(res.token);
    return res.token;
  } catch (e) {
    throw e;
  }
}
```

mostly for stylistic purposes so users can do:
```
const { handleLogin } = useSession()
handleLogin();
```

instead of 
```
const { loginMutation } = useSession()
loginMutation.mutate()
```

Reply:
I just tried out this pattern but haven't had much success with it. It's possible that I'm not using `useMutation` correctly, but it seems like `await` on the mutation may return before the `onSuccess` callback, which makes it hard to trigger the redirect after the auth token is set. I wonder how this is done in playground?

Reply:
Added a ticket to address this with a new PR